File: pset6/sentimental-readability/readability.py
"""
readability.py asks users to input text and gives a readability score based off the Coleman-Liau index.
"""

from cs50 import get_string
import re


def main():

    # Gets text from the user
    text = get_string("Text: ")

    # Functions to get the letter, word, and sentences
    l = letter_count(text)
    w = word_count(text)
    s = sentence_count(text)

    average_letters = (l / w) * 100
    average_sentences = (s / w) * 100

    # Computing the Coleman-Liau index
    reading_grade = int(round((.0588 * average_letters) - (.296 * average_sentences) - 15.8))

    #reading_grade = int(reading_index)

    # Uses conditionals to print out reading score.
    if reading_grade < 1:
        print("Before Grade 1")
    elif reading_grade >= 16:
        print("Grade 16+")
    else:
        print(f"Grade {reading_grade}")


def letter_count(words):
    """Gets the average letters per 100 words"""
    # Creating a regex object to search for letters
    letter_regex = re.compile(r"[a-zA-Z]")

    letters = 0

    # Iterates through every word and every character to see if it is a valid letter
    for characters in words:
        letter_search = letter_regex.search(characters)
        if letter_search != None:
            letters += 1

    return letters


def word_count(words):
    """Takes in a string to count the number of words within that string."""

    list_of_words = words.split()
    number_of_words = len(list_of_words)
    return number_of_words


def sentence_count(words):
    """Returns the number of sentences there are in a string."""

    punctuation = (".", "?", "!")
    sentences = 0
    for character in words:
        if character in punctuation:
            sentences += 1

    return sentences


if __name__ == "__main__":
    main()

File: pset2/readability.c
#include <cs50.h>
#include <ctype.h>
#include <math.h>
#include <stdio.h>
#include <string.h>

float count_letters(string text);
float count_words(string text);
float count_sentences(string text);
int calculate_index(float words, float sentences, float letters);

int main(void)
{
    string text = get_string("Text: ");
    float letters = count_letters(text);
    float words = count_words(text);
    float sentences = count_sentences(text);
    int index = calculate_index(words, sentences, letters);
    if (index > 0 && index <= 16)
    {
        printf("Grade %i\n", index);
    }
    else if (index < 1)
    {
        printf("Before Grade 1\n");
    }
    else
    {
        printf("Grade 16+\n");
    }
}

float count_letters(string text)
{
    float letters = 0;
    for (int i = 0, n = strlen(text); i < n; i++)
    {
        char c = toupper(text[i]);
        if (c >= 65 && c <= 90)
        {
            letters += 1;
        }
    }
    return letters;
}

float count_words(string text)
{
    float words = 1;
    for (int i = 0, n = strlen(text); i < n; i++)
    {
        if (text[i] == ' ')
        {
            words += 1;
        }
    }
    return words;
}

float count_sentences(string text)
{
    float sentences = 0;
    for (int i = 0, n = strlen(text); i < n; i++)
    {
        char c = text[i];
        if (c == '.' || c == '?' || c == '!')
        {
            sentences += 1;
        }
    }
    return sentences;
}

int calculate_index(float words, float sentences, float letters)
{
    float avg_letters = (letters / words) * 100;
    float avg_sentences = (sentences / words) * 100;
    float index = (.0588 * avg_letters) - (.296 * avg_sentences) - 15.8;
    int i = round(index);
    return i;
}
